util/podprotector/multi_informer.go

```diff
@@ -0,0 +1,403 @@
+	Watch(ctx context.Context) <-chan map[SourceName]SourceDesc
+
+	// Updates the status of a PodProtector object as received from `sourceName`.
+	UpdateStatus(ctx context.Context, sourceName SourceName, ppr *podseidonv1a1.PodProtector) error
+}
+
+// Frequently used type in component args.
+// Requests the relevant dependencies for a source provider,
+// and returns the getter for the actual source provider that can be accessed after init.
+// The return type is `func() SourceProvider` instead of `component.Dep[SourceProvider]`
+// to support use cases where the actual component does not use `SourceProvider` as its API directly.
+type SourceProviderRequest = func(*component.DepRequests) func() SourceProvider
+
+// Identifies a source for PodProtector.
+type SourceName string
+
+// Describes how to access a source cluster.
+type SourceDesc struct {
+	// A client to access the PodProtectors in the cluster.
+	//
+	// Only List and Watch methods of the interface are used.
+	// Updates will go through [SourceProvider.UpdateStatus] instead.
+	PodseidonClient podseidonclient.Interface
+}
+
+type informerState struct {
+	isSourceIdent bool
+	sources       atomic.Pointer[map[SourceName]*sourceState]
+	postHandlers  []func(PodProtectorKey)
+}
+
+type sourceState struct {
+	lister        podseidonv1a1listers.PodProtectorLister
+	hasSynced     cache.InformerSynced
+	selectorIndex SelectorIndex
+	cancelFunc    context.CancelFunc
+}
+
+func (state *informerState) updateSourceList(
+	ctx context.Context,
+	cellList map[SourceName]SourceDesc,
+	obs observer.IndexedInformerObserver,
+) {
+	if err := state.tryUpdateSourceList(ctx, cellList, obs); err != nil {
+		obs.UpdateSourceListError(ctx, observer.UpdateSourceListError{Err: err})
+	}
+}
+
+func (state *informerState) tryUpdateSourceList(
+	ctx context.Context,
+	newDescs map[SourceName]SourceDesc,
+	obs observer.IndexedInformerObserver,
+) error {
+	prevMapPtr := state.sources.Load()
+	prevMap := ptr.Deref(prevMapPtr, nil)
+
+	nextMap := map[SourceName]*sourceState{}
+	uncommitted := []context.CancelFunc{}
+
+	defer func() {
+		for _, cancelFunc := range uncommitted {
+			cancelFunc()
+		}
+	}()
+
+	for name, desc := range newDescs {
+		sourceState, hasName := prevMap[name]
+		if !hasName {
+			newSourceState, err := startSourceInformer(ctx, obs, name, desc, state.postHandlers)
+			if err != nil {
+				return err
+			}
+
+			sourceState = newSourceState
+			uncommitted = append(uncommitted, newSourceState.cancelFunc)
+		}
+
+		nextMap[name] = sourceState
+	}
+
+	additions := len(uncommitted)
+
+	{
+		// atomically uncommit prevMap and nextMap committed
+		// thus, the uncommitted list changes from (nextMap - prevMap) to (prevMap - nextMap).
+		swapped := state.sources.CompareAndSwap(prevMapPtr, ptr.To(nextMap))
+
+		if !swapped {
+			panic("unexpected informer map pointer write from another goroutine")
```

Comment:
Is the `swapped fail` a cannot recovery issue? Could we try to fix it when we do the next round reconcile?

Reply:
This is an unexpected condition, practically a race that we don't know where this came from, since this is the only goroutine. I prefer to panic until we figure out when this happens.